src/catalog/dependency_manager.cpp

```diff
@@ -27,166 +115,322 @@ void DependencyManager::AddObject(CatalogTransaction transaction, CatalogEntry &
 		if (!dependency.set) {
 			throw InternalException("Dependency has no set");
 		}
-		auto catalog_entry = dependency.set->GetEntryInternal(transaction, dependency.name, nullptr);
+		auto catalog_entry = dependency.set->GetEntry(transaction, dependency.name);
 		if (!catalog_entry) {
 			throw InternalException("Dependency has already been deleted?");
 		}
 	}
+
 	// indexes do not require CASCADE to be dropped, they are simply always dropped along with the table
 	auto dependency_type = object.type == CatalogType::INDEX_ENTRY ? DependencyType::DEPENDENCY_AUTOMATIC
 	                                                               : DependencyType::DEPENDENCY_REGULAR;
 	// add the object to the dependents_map of each object that it depends on
 	for (auto &dependency : dependencies.set) {
-		auto &set = dependents_map[dependency];
-		set.insert(Dependency(object, dependency_type));
+		auto &dependency_connections = GetOrCreateDependencySet(transaction, dependency);
+		dependency_connections.AddDependent(transaction, object, dependency_type);
 	}
 	// create the dependents map for this object: it starts out empty
-	dependents_map[object] = dependency_set_t();
-	dependencies_map[object] = dependencies.set;
+
+	auto &object_connections = GetOrCreateDependencySet(transaction, object);
+	object_connections.AddDependencies(transaction, dependencies);
+}
+
+static bool CascadeDrop(bool cascade, DependencyType dependency_type) {
+	if (cascade) {
+		return true;
+	}
+	if (dependency_type == DependencyType::DEPENDENCY_AUTOMATIC) {
+		// These dependencies are automatically dropped implicitly
+		return true;
+	}
+	if (dependency_type == DependencyType::DEPENDENCY_OWNS) {
+		// The object has explicit ownership over the dependency
+		return true;
+	}
+	return false;
+}
+
+void DependencyManager::UnmangleName(const string &mangled, CatalogType &type, string &schema, string &name) {
+	auto parts = StringUtil::Split(mangled, std::string("\0", 1));
+	D_ASSERT(parts.size() == 3);
+	type = CatalogTypeFromString(parts[0]);
+	schema = std::move(parts[1]);
+	name = std::move(parts[2]);
+}
+
+void GetLookupProperties(CatalogEntry &entry, string &schema, string &name, CatalogType &type) {
+	if (entry.type == CatalogType::DEPENDENCY_ENTRY) {
+		auto &dependency_entry = entry.Cast<DependencyCatalogEntry>();
+
+		schema = dependency_entry.EntrySchema();
+		name = dependency_entry.EntryName();
+		type = dependency_entry.EntryType();
+	} else if (entry.type == CatalogType::DEPENDENCY_SET) {
+		auto &dependency_set = entry.Cast<DependencySetCatalogEntry>();
+
+		auto &mangled_name = dependency_set.MangledName();
+		DependencyManager::UnmangleName(mangled_name, type, schema, name);
```

Comment:
We can probably get rid of UnmangleName entirely if we just store the `schema` `name` and `type` in the DependencySetCatalogEntry as well